runtime/autoload/man.vim

```diff
@@ -2,14 +2,64 @@
 
 let s:man_find_arg = "-w"
 
+" Handler for s:system() function.
+function! s:system_handler(jobid, data, event) dict abort
+  if a:event == 'stdout'
+    let self.stdout .= join(a:data, "\n")
+  elseif a:event == 'stderr'
+    let self.stderr .= join(a:data, "\n")
+  else
+    let self.exit_code = a:data
+  endif
+endfunction
+
+" Run a system command and timeout after 30 seconds.
+function! s:system(cmd, ...) abort
+  let opts = {
+        \ 'stdout': '',
+        \ 'stderr': '',
+        \ 'exit_code': 0,
+        \ 'on_stdout': function('s:system_handler'),
+        \ 'on_stderr': function('s:system_handler'),
+        \ 'on_exit': function('s:system_handler'),
+        \ }
+  let jobid = jobstart(a:cmd, opts)
+
+  if jobid < 1
+    throw printf('command error %d: %s', jobid, join(a:cmd))
+  endif
+
+  let res = jobwait([jobid], 30000)
+  if res[0] == -1
+    try
+      call jobstop(jobid)
+      throw printf('command timed out: %s', join(a:cmd))
+    catch /^Vim\%((\a\+)\)\=:E900/
+    endtry
+  elseif res[0] == -2
+    throw printf('command interrupted: %s', join(a:cmd))
+  endif
+  if opts.exit_code != 0
+    throw printf("command error (%d) %s: %s", jobid, join(a:cmd), substitute(opts.stderr, '\_s\+$', '', &gdefault ? '' : 'g'))
+  endif
+
+  return opts.stdout
+endfunction
+
 " TODO(nhooyr) Completion may work on SunOS; I'm not sure if `man -l` displays
 " the list of searched directories.
 try
   if !has('win32') && $OSTYPE !~? 'cygwin\|linux' && system('uname -s') =~? 'SunOS' && system('uname -r') =~# '^5'
     let s:man_find_arg = '-l'
   endif
+  let s:man_has_l_flag = 1
+  let man_file = s:system(['man', '-w', 'man'])
+  call s:system(['env', 'MANPAGER=cat', 'man', '-l', man_file])
+  unlet man_file
```

Comment:
Yes that seems reasonable. The thing is, I'm not a big fan of the `s:system` solution since I'm not quite sure if another exception can get thrown somewhere in there and I'll might end up with 'man_has_l_flag' as a default. I'd prefer the `system` solution.